### DISCUSSION TITLE
Gradient Clipping

### DISCUSSION THREAD

**Discussion by anais2390** [Project-MONAI/MONAI#3892] [Q&A]
Hi everyone, 

Could you tell the correct way to perform gradient clipping with monai and pytorch ignite ? 
Thanks a lot !

**Comment by Nic-Ma** [comment]
Hi @anais2390 ,

MONAI provides rich components for PyTorch training program, so actually, I think you can do `gradient clipping` just like a regular PyTorch program.
And if you are using `ignite`, I think @vfdev-5 may help provide some examples?

Thanks in advance.

**Reply by anais2390** [reply]
Thanks you for your answer. I could really use an example using pytorch ignite, I can't figure out how to do it. Thanks a lot for your help.

**Reply by vfdev-5** [reply]
@Nic-Ma thanks for pinging !

@anais2390 to do gradient clipping usually it is done inside a training step. For example, for a supervised learning single model:
```python
from torch.nn.utils import clip_grad_norm_
from ignite.engine import Engine

model = ...
optimizer = ...
criterion = ...
gradient_clip = ...
train_loader = ...

def train_step(engine, batch):
    model.train()
    optimizer.zero_grad()
    x, y = batch[0].to(device), batch[1].to(device)
    y_pred = model(x)
    loss = criterion(y_pred, y)
    loss.backward()

    # Here is grad clipping:
    clip_grad_norm_(model.parameters(), gradient_clip)

    optimizer.step()
    return loss.item()

trainer = Engine(train_step)

trainer.run(train_loader, max_epochs=100)
```

Let me know if it helps

**Reply by anais2390** [reply]
Hi @vfdev-5 and thank you for you answer !

Our problem is that we use SupervisedTrainer from monai. 
Is it possible to specified to the SupervisedTrainer that we want to use gradient clipping ? Or do we need to create our own custom Trainer to be able to define a gradient clipping ? 

Thanks again for your help !

**Reply by Nic-Ma** [reply]
Hi @anais2390 ,

I think you can define your own `iteration_update` arg in `SupervisedTrainer`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/engines/trainer.py#L81

Thanks.

**Reply by anais2390** [reply]
Hi @Nic-Ma, 

Thanks you, that is exactly what I am looking for !

Could anyone give me an example on how to define properly gradient clipping using this argument `iteration_update` in `SuperviedTrainer` ?

Thanks again !

**Reply by Nic-Ma** [reply]
You can refer to this code:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/apps/deepgrow/interaction.py#L55

Thanks.

**Reply by vfdev-5** [reply]
@anais2390 I think you can add a `grad_clip_handler` handler to `IterationEvents.BACKWARD_COMPLETED` event (https://github.com/Project-MONAI/MONAI/blob/4b66746764e200a9a933d2198a7aee13e53b1206/monai/engines/trainer.py#L212)
```python
from torch.nn.utils import clip_grad_norm_
from monai.engines.trainer import SupervisedTrainer, IterationEvents

trainer = SupervisedTrainer(...)

gradient_clip = 0.1

@trainer.on(IterationEvents.BACKWARD_COMPLETED)
def grad_clip_handler():
    clip_grad_norm_(self.network.parameters(), gradient_clip)
```

@Nic-Ma what do you think ?

**Reply by Nic-Ma** [reply]
Yeah, that’s also good and may be easier.

Thanks @vfdev-5

**Reply by anais2390** [reply]
Hi @Nic-Ma  and @vfdev-5,

Thanks a lot for your help !
I had actually looked at the method through the `IterationEvents`, but I thought that the gradient clipping should be done before the backpropagation. Are you sure it is ok to do it on the event `BACKWARD_COMPLETED` ?

**Reply by vfdev-5** [reply]
@anais2390 please check the code: https://github.com/Project-MONAI/MONAI/blob/4b66746764e200a9a933d2198a7aee13e53b1206/monai/engines/trainer.py#L212 and you will see that `IterationEvents.BACKWARD_COMPLETED` is triggered after computing gradients, after `loss.backward()` but before applying them to the parameters: `optimizer.step()`.

[Here](https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping) is another example in pytorch docs how to apply grad clipping with AMP. You can ignore AMP part and see that the steps are
- loss.backward
- grad clipping
- optimizer.step

Let me know if I misunderstood your question. Thanks

**Reply by anais2390** [reply]
This is very clear, thanks a lot for your help !
